### ISSUE TITLE
BatchNormNd layer

### ISSUE THREAD

**Issue by davidmascharka** [davidmascharka/MyNN#29]
We should have one. [Here's a link](https://arxiv.org/abs/1502.03167) to the original paper.

**Comment by IanCoolidge0** [comment]
I've been working on this layer slightly over the past couple of weeks, and it is currently present in the batch_norm branch on my fork. I plan on making a pull request once it works baseline for 1-dimensional batch norm (I'm testing with the spiral dataset).

A few problems I have run into:

1. Should this layer be able to perform an n-dimensional batch norm for any n, or just 2 and 4 dimensions (for dense layers and 2d conv layers respectively)? Right now my layer just does 2 and 4, however I don't think it would be too difficult to generalize this (just a matter of figuring out which axes to normalize over).

2. What is the best way to calculate the moving averages for testing? Obviously it isn't stored as a mygrad tensor since it's not a learnable parameter, but I wasn't sure what the best way was to calculate it. Currently my layer does it by storing a count of the number of training forward passes, and uses this counter to calculate any mean from the previous mean. This seemed like an odd way to do it though, so if there's a better way I would definitely implement that.

**Comment by davidmascharka** [comment]
Awesome! This will be a great layer to have access to.

1. Ideally, this should be n-dimensional to match the pooling and convolutional layers. It's a bit awkward to have `conv_nd`, `maxpool_nd`, and then `batchnorm_1d`, `batchnorm_2d`, and `batchnorm_3d`. The extra work involved in computing dimensions over which to compute statistics will pay dividends with the simplicity of the interface.

2. The running mean and std can be be computed using an exponential moving average. This should be:

```
running_stat = (1 - momentum)*running_stat + momentum*batch_stat
``` 

Here, momentum determines the past vs current weighting (0.1 is common but this should be a parameter in the `__init__`), `running_stat` is what we're keeping track of, and `batch_stat` is our batch statistic.

**Comment by rsokl** [comment]
@IanCoolidge0 let me know if you would like some input about generalizing the math to accommodate multiple dimensions and/or with deriving the gradient.

Once you have a PR up, we can discuss how to handle the running statistics - I have experience with computing running statistics in this sort of way.

**Comment by rsokl** [comment]
Great!

**Comment by IanCoolidge0** [comment]
Okay, I put the PR up. If the dimensions don't generalize as easily as they seem to (just calculate over axis 0 or whatever the batch axis is regardless?) I think that would be the main thing that would still need to be fixed. Thanks!

**Comment by rsokl** [comment]
@IanCoolidge0  nice work 🎆
